core/src/main/scala/filodb.core/memstore/OnDemandPagingShard.scala

```diff
@@ -140,7 +140,12 @@ TimeSeriesShard(dataset, storeConfig, shardNum, rawStore, metastore, evictionPol
                 unsafeKeyOffset = PartKeyLuceneIndex.bytesRefToUnsafeOffset(partKeyBytesRef.offset)
                 group = partKeyGroup(dataset.partKeySchema, partKeyBytesRef.bytes, unsafeKeyOffset, numGroups)
                 part <- Option(createNewPartition(partKeyBytesRef.bytes, unsafeKeyOffset, group, 4)) } yield {
-            partSet.add(part)
+            val stamp = partSetLock.writeLock()
```

Comment:
The write lock acquisition isn't a concern. Each loop iteration calls into Lucene and creates a new partition. I cannot infer the cost of these operations because to some extent, they're a black box. In fact, creating a new partition allocates memory, which might in turn block some more. So holding the partSetLock for a longer time stalls queries which need to read from partSet.

The usual trick I apply in cases like this doesn't work with a StampedLock:  Acquire the lock outside the loop and keep it. In the middle of the loop, check if there's a queue of waiters to gain access to the lock. If so, release and re-acquire to let them through. There's no method on StampedLock to ask if there's a queue.